src/redux/modules/spinner/spinner.js

```diff
@@ -0,0 +1,97 @@
+/* @flow */
+import debug from 'debug';
+import {
+  UPDATE_USER_REQUEST,
+  UPDATE_USER_SUCCESS,
+  UPDATE_USER_FAILURE,
+} from 'redux/modules/user/user-actions';
+import {
+  LOGIN_REQUEST,
+  LOGIN_SUCCESS,
+  LOGIN_FAILURE,
+  LOGOUT_REQUEST,
+  LOGOUT_SUCCESS,
+} from 'redux/modules/auth/auth-actions';
+
+type SpinnerState = {
+  canShow: boolean;
+  messageId: string;
+};
+
+export const DEFAULT_SPINNER_STATE: SpinnerState = {
+  canShow: false,
+  messageId: 'please_wait',
+};
+
+if (__DEBUG__) {
+  debug.enable('spinner-reducer:*');
+}
+
+const log = debug('spinner-reducer:debug');
+
+const spinnerReducer = (state : SpinnerState = DEFAULT_SPINNER_STATE, action: any): any => {
+  let newState: SpinnerState;
+
+  switch (action.type) {
+    case UPDATE_USER_REQUEST:
+      newState = Object.assign(
+        {},
+        action.state,
+        {
+          canShow: true,
+          messageId: 'updating_user_details',
+        }
+      );
+      break;
+    case UPDATE_USER_SUCCESS:
+      newState = Object.assign({}, action.state, DEFAULT_SPINNER_STATE);
+      break;
+    case UPDATE_USER_FAILURE:
+      newState = Object.assign({}, action.state, DEFAULT_SPINNER_STATE);
+      break;
+    case LOGIN_REQUEST:
+      // TODO: i18n for `message`
+      // TODO: where should the value for `message` be set?
+      newState = Object.assign(
+        {},
+        action.state,
+        {
+          canShow: true,
+          messageId: 'logging_in',
+        }
+      );
+      break;
+    case LOGIN_SUCCESS:
+      newState = Object.assign({}, action.state, DEFAULT_SPINNER_STATE);
+      break;
+    case LOGIN_FAILURE:
+      newState = Object.assign({}, action.state, DEFAULT_SPINNER_STATE);
+      break;
+    case LOGOUT_REQUEST:
+      // TODO: i18n for `message`
+      // TODO: where should the value for `message` be set?
+      newState = Object.assign(
+        {},
+        action.state,
+        {
+          canShow: true,
+          messageId: 'logging_out',
+        }
+      );
+      break;
+    case LOGOUT_SUCCESS:
+      newState = Object.assign({}, action.state, DEFAULT_SPINNER_STATE);
+      break;
+    default:
+      newState = state;
```

Comment:
Since spinners _always_ happen during async actions, I see the action creators as a candidate. In reality, when we do a `LOGIN_REQUEST` action, we only do it to update a loading state of some sort (although a local one, perhaps just used in a login widget). It makes sense to dispatch a global  loading state action at the same time. In `loginRequest`, at the same time you dispatch the `LOGIN_REQUEST` action, you dispatch another action: https://github.com/mcalthrop/opencredo-react-boilerplate-1/blob/feature/32-page-with-data-from-server/src/redux/modules/auth/auth-actions.js#L78-L97

```
    dispatch(showSpinner(message)); // <-- show your spinner here
    dispatch({
      type: LOGIN_REQUEST,
      state: {
        isLoading: true,
        isAuthenticated: false,
      },
    });

    ...
    // do your api call and handle the promise
    .then(response => {
      dispatch(loginSuccess());
      dispatch(hideSpinner()); // <-- hide your spinner here
    }, failure => {
      dispatch(loginFailure()); // <-- hide your spinner here
      dispatch(hideSpinner());
    });
```

That way, actionCreators are _declarative_ about what's going to happen in the application.